Feature/First iteration on MC failsafe

anton-matosov opened the pull request:
Add failsafe tune
Fix LED blinking for Pixracer 
Pass rc_loss_mode instead of simple bool flag
Add setting of failsafe PWM outputs in case of force_failsafe

LorenzMeier commented:
Nice!

LorenzMeier commented:
@kd0aij Can you test?

kd0aij commented:
What is rc_loss_mode supposed to do? I guess this should fix the flight termination bug I mentioned in the RC loss PR... need to reference it from here.

kd0aij commented:
what about #5388?

anton-matosov commented:
Rc_loss_mode is enum that's is not used right now, but it should define behavior of RC loss failsafe action. With current code it switches to "land" mode, and there is no way to force termination
I took a brief look at your PR and it does some of what I am missing right now regarding applying failsafe during termination. But it also does some stuff a bit brute force (e.g. applying lockdown from commander state machine)
I will incorporate your fixes into my PR, so it will be a complete solution (right now it drops output to minimum pwm upon termination)

julianoes commented on src/modules/commander/commander.cpp:
```diff
@@ -2881,7 +2881,7 @@ int commander_thread_main(int argc, char *argv[])
 						       &status_flags,
 						       land_detector.landed,
-						       (rc_loss_enabled > 0),
+						       rc_loss_mode,
```
I like this change but does it change the behaviour at all like this?

julianoes replied:
Note that the same param is used again in the navigator which is really quite ugly:
https://github.com/PX4/Firmware/blob/master/src/modules/commander/commander.cpp#L1276
https://github.com/PX4/Firmware/blob/master/src/modules/navigator/navigator_main.cpp#L162

anton-matosov replied:
No, it doesn't, yet.
I have never seen Navigator before. What is it used for?
It looks like it has very similar logic as commander for data/rc links and GPS, but handles it in its own way.

anton-matosov replied:
I read more commander and navigator code and now feel the difference of their responsibilities. 
(Please correct me if I am wrong)
Navigator simply performs 'pieces' of the autonomous navigation, it doesn't make a decision.
Commander is that one that maintains the global state of the vehicle based on params, user and sensor inputs and makes all the decisions of what action to take in particular situation.

If my understanding is correct then I am 100% agree that rc_loss_mode should be handled in commander (SM helper), not navigator.

My only concern is that SM helper keeps growing and already has a big chunk of copy-pasted code. I feel a bit worried of breaking things by doing change to RC loss like this without having really good understanding and having unit tests covering the logic.

julianoes replied:
@anton-matosov I think you got it right, that's very much the way it was designed. If you're concerned about breaking the state machine helper the way to resolve that would be to add tests to and then extend or refactor it.

anton-matosov commented:
Ok. I think that's what I am going to do then. I will write tests at least for code I am going to modify for SM Helper and navigator. Are there any existing tests?

dagar commented:
@anton-matosov state machine tests here - https://github.com/PX4/Firmware/blob/master/src/modules/commander/commander_tests/state_machine_helper_test.cpp

Nothing really for navigator, unless you can fit it into a mission test. https://github.com/PX4/Firmware/tree/4453e4201b7a245cff52beeb38a293161aea4c48/integrationtests/python_src/px4_it/mavros

AndreasAntener commented:
@anton-matosov what do you plan to change in the state machine helper? Because I'm reworking this at the moment and implemented a different way of how the failures are handled (more comprehensive), also adding unit tests.

anton-matosov commented:
@AndreasAntener I am planning to refactor RC and Datalink loss code to get rid of the copy-paste and migrate RC/Data loss mode switching to commander out from Navigator the same way offboard loss modes are handled.
Also I will extend loss mode to force termination mode. (that is what I need for FPV racers)

anton-matosov commented:
@AndreasAntener I am planning to refactor RC and Datalink loss code to get rid of the copy-paste and migrate RC/Data loss mode switching to commander out from Navigator the same way offboard loss modes are handled.
Also I will extend loss mode to force termination mode. (that is what I need for FPV racers)

anton-matosov commented:
@AndreasAntener are you done with your refactoring of the commander's SM?
Which fork/branch are you working in? (so I can merge in your changes and keep going)

AndreasAntener commented:
@anton-matosov I think we should separate the refactoring from your functionality change (termination handling). I'll push a branch in a sec that contains the move from navigator, after I can build upon that again.

AndreasAntener commented:
@anton-matosov added a PR for you (https://github.com/PX4/Firmware/pull/5884), please have a look. You integrate the changes from there into your branch and delete it, or whatever is easier for you

anton-matosov commented:
@AndreasAntener @LorenzMeier @kd0aij @tumbili @julianoes PR is ready for review

LorenzMeier commented:
Thanks! I did a first pass and it overall looks good. A couple of questions:

  * What was the driver behind adding `manual_lockdown` in addition to the `lockdown` flag? Was it necessary to have both?
  * Do you require manual RC input in SITL? I saw you added it back?
  * Can you explain more about the build system changes? They look ok, but I'd like to understand the use case.

Please try to describe this in general when submitting a PR, since those are changes that need context. Thanks!

anton-matosov commented:
> What was the driver behind adding manual_lockdown in addition to the lockdown flag? Was it necessary to have both?

If you have kill switch configured on RC, `lockdown` will be constantly reset to the killswitch position and it can't be used from other code. 

> Do you require manual RC input in SITL? I saw you added it back?

I was using manual RC for the RC loss testing and I made it off by default (configurable in CMakeLists). I will also need it for RC expo rates code testing in simulator.

>Can you explain more about the build system changes? They look ok, but I'd like to understand the use case.

I have added wrapper for px4 that includes paths originally provided via the command line by `cmake`+`sitl_run.sh`. The original approach was not allowing easy debugging from IDE (I use CLion). With current approach you can simply chose simulator target you want to debug, e.g. `px4_iris` and simply hit `debug` button and it will build and start debug session with no need to configure bunch of run parameters. 
I have also added `IDE` parameter to `debuggers` which is handled in a way of skipping launching `px4` from `sitl_run.sh` that allows simple start/stop of `px4` debugging from IDE without the need to restart viewer (I was using `gazebo`)

anton-matosov commented:
@LorenzMeier any updates?

julianoes commented:
@anton-matosov will review this weekend, thanks for your work.

anton-matosov commented:
Thanks @julianoes

julianoes commented on src/modules/commander/state_machine_helper.h:
```diff
@@ -58,6 +58,16 @@ typedef enum {
+	AUTO_LAND = 3,
+	AUTO_RECOVER = 4,
+	TERMINATE = 5,
+	LOCKDOWN = 6,
```
Wouldn't `TERMINATE` and `LOCKDOWN` be the same?

anton-matosov replied:
Currently flight termination is implemented differently for fixed wing. For MC it will be pretty much the same. But in the future Terminate can for example open parachute, or trigger some other flight termination procedure, while lockdown is always just motors lockdown

julianoes commented on src/modules/commander/state_machine_helper.h:
```diff
@@ -58,6 +58,16 @@ typedef enum {
+	AUTO_LOITER = 1,
+	AUTO_RTL = 2,
+	AUTO_LAND = 3,
+	AUTO_RECOVER = 4,
```
What is `AUTO_RECOVER`? I wonder if that is (still) functional.

anton-matosov replied:
It is from existing code, I just gave it a name based on existing param description.

julianoes commented on src/modules/mc_att_control/mc_att_control_main.cpp:
```diff
@@ -1040,6 +1040,63 @@ MulticopterAttitudeControl::task_main()
 				}
 			}
+
+			if (_v_control_mode.flag_control_termination_enabled) {
```
Shouldn't this be taken care off in the drivers instead of the controllers?

anton-matosov replied:
This where I have stared based on the FW_att_control but I couldn't stop motors from here. 
For lockdown we don't really need anything, but for termination it is the a kind of only place to do something.
 IMO drivers should not know about flight control modes and take actions based on this.

LorenzMeier replied:
@anton-matosov The whole diff / PR makes sense, this is the only section I don't get yet entirely.

Are you trying to prevent an integrator windup here or is this needed to stop the props?

julianoes commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -565,15 +563,7 @@ Navigator::task_main()
-				} else { /* if == 2 or unknown, RTL */
-					_navigation_mode = &_rtl;
-				}
+				_navigation_mode = &_rcLoss;
```
This is much cleaner and moves the responsibility to the commander. I think that's the right approach.

anton-matosov replied:
Cudos to @AndreasAntener

julianoes commented on src/modules/simulator/simulator_config.h.in:
```diff
@@ -0,0 +1,39 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2016 Anton Matosov. All rights reserved.
```
This should be PX4 Development Team, I think.

anton-matosov replied:
Emmm... I am not part of that team, so I don't think I should put your name on my changes.

julianoes replied:
It's just the default header in general, not sure.

julianoes commented on src/modules/simulator/simulator_mavlink.cpp:
```diff
@@ -1,6 +1,7 @@
 /****************************************************************************
  *
  *   Copyright (c) 2015 Mark Charlebois. All rights reserved.
+ *   Copyright (c) 2016 Anton Matosov. All rights reserved.
```
And the same here actually.

julianoes commented on src/platforms/posix/sitl_runner_main.cpp.in:
```diff
@@ -0,0 +1,53 @@
+        return errno;
+    }
+    const char *argsOverride[] = {argv[0], "@SITL_RUNNER_SOURCE_DIR@", "@SITL_RUNNER_MODEL_FILE@"};
+    return SITL_MAIN(sizeof(argsOverride) / sizeof(argsOverride[0]), (char**)argsOverride);
```
I'm gonna push back on this change, I think it conflicts heavily with upcoming changes in #5162.

anton-matosov replied:
I don't see any conflicts with your PR (on logic/behavior level, didn't dive in code yet)
But to make it cleaner I will separate build system/debugging changes from failsafe code and submit them as separate PR

julianoes commented on src/modules/commander/state_machine_helper.cpp:
```diff
@@ -1102,6 +1060,93 @@ bool set_nav_state(struct vehicle_status_s *status, struct commander_state_s *in
+	set_link_loss_nav_state(status, armed, status_flags, link_loss_act, vehicle_status_s::NAVIGATION_STATE_AUTO_RTGS);
+}
+
+void set_link_loss_nav_state(struct vehicle_status_s* status,
```
Oh I realize now what you are doing here. It seems quite complicated, in my opinion we need to simplify this all.

anton-matosov replied:
What exactly looks complicated here? This is mostly port of the link loss code from the navigator to commander + 2 new action types.
I have introduced enum for actions instead of using magic numbers and merged handlers for RC and data loss as the are the same except 1 nav state.

julianoes replied:
Got it. Were you already able to test most/all the cases?

anton-matosov replied:
I have not tested all the cases.

anton-matosov commented on src/drivers/pwm_out_sim/pwm_out_sim.cpp:
```diff
@@ -512,6 +516,19 @@ PWMSim::task_main()
+			/* overwrite outputs in case of force_failsafe */
+			if (_failsafe) {
+				for (size_t i = 0; i < num_outputs; i++) {
+					outputs.output[i] = 600;
```
I chose 600 as a magic for PWM Simulator failsafe as it doesn't read values from settings (might need to add this instead) and I needed some distinguishable value from disarmed to easily debug failsafe and not confuse with other states.

anton-matosov replied:
I have moved 600 and 900 to named constants.
It doesn't make sense to read params as they are not configured for any simulator

anton-matosov commented on Tools/sitl_run.sh:
```diff
@@ -135,6 +135,15 @@ then
 elif [ "$debugger" == "valgrind" ]
 then
 	valgrind $sitl_command
+elif [ "$debugger" == "ide" ]
```
Any suggestions on where to put docs for this?

anton-matosov commented on src/modules/commander/commander.cpp:
```diff
@@ -3170,19 +3173,28 @@ control_status_leds(vehicle_status_s *status_local, const actuator_armed_s *actu
+			}
+		} else {
+			/* armed, solid */
+			led_on(LED_BLUE);
```
Most likely you are right. I will double check and fix. Good catch!

anton-matosov commented on src/modules/commander/commander.cpp:
```diff
@@ -2915,18 +2915,19 @@ int commander_thread_main(int argc, char *argv[])
-						       offboard_loss_act,
-						       offboard_loss_rc_act);
+        /* now set navigation state according to failsafe and main state */
+        bool nav_state_changed = set_nav_state(&status,
```
Ok, will fix that

LorenzMeier replied:
Everything has to be tabs and default is 8 width (if you use Sublime, you get a decent default). Anything having mixed formatting is the result of style checking not being enforced yet (e.g. in the commander). We're cleaning up module-by-module and probably should enforce style when this PR is in.

anton-matosov commented on src/modules/commander/state_machine_helper.cpp:
```diff
@@ -117,15 +117,26 @@ static const char *const state_names[vehicle_status_s::ARMING_STATE_MAX] = {
 static hrt_abstime last_preflight_check = 0;	///< initialize so it gets checked immediately
 static int last_prearm_ret = 1;			///< initialize to fail
 
+void set_link_loss_nav_state(struct vehicle_status_s* status,
```
`make format` doesn't catch this.

anton-matosov replied:
fixed in all over the state_machine_helper

anton-matosov commented:
@julianoes can you please take a look at the latest changes that address your comments?

anton-matosov commented:
@LorenzMeier @julianoes I have applied suggested changes. Please review again.

LorenzMeier commented:
@anton-matosov Thanks, much appreciated! I'm going to try to get this in tomorrow.

LorenzMeier commented:
Awesome! This is ready to go, one last question. @anton-matosov Great work!

LorenzMeier commented:
Rebased and continuing here:
https://github.com/PX4/Firmware/pull/6177
